{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mini Project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[nltk_data] Downloading package stopwords to\n",
      "[nltk_data]     C:\\Users\\jmachalek\\AppData\\Roaming\\nltk_data...\n",
      "[nltk_data]   Package stopwords is already up-to-date!\n"
     ]
    }
   ],
   "source": [
    "import logging\n",
    "\n",
    "from storage_functions import create_astra_vstore, load_pdf, chunk_docs, load_docs_from_db\n",
    "from eval_functions import make_test_set, add_llm_answers_to_dataset, evaluate_dataset\n",
    "from explain_functions import format_answer\n",
    "from llm_functions import perform_rag\n",
    "\n",
    "# Embedding Models\n",
    "from langchain_openai import OpenAIEmbeddings\n",
    "\n",
    "# LLM Models\n",
    "from langchain_openai import ChatOpenAI\n",
    "\n",
    "# Explainability\n",
    "from IPython.display import HTML\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:logging info\n"
     ]
    }
   ],
   "source": [
    "logging.basicConfig(level=logging.WARNING)\n",
    "logging.info(\"logging info\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Predeclare values\n",
    "file_path = \"aha_hcm_guidance_2020.pdf\"\n",
    "emb_model = \"text-embedding-3-small\"\n",
    "openai_embedding = OpenAIEmbeddings(model = emb_model)\n",
    "llm = ChatOpenAI(model = \"gpt-3.5-turbo\")\n",
    "collection_name = \"aha_guidelines\"\n",
    "score_threshold = 0.7\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def vstore_pipeline():\n",
    "    \"\"\"\n",
    "    Performs a pipeline of operations to create and populate an Astra VStore.\n",
    "\n",
    "    Returns:\n",
    "        astra_vstore (AstraVStore): The created Astra VStore object.\n",
    "        docs (list): The list of loaded PDF documents.\n",
    "    \"\"\"\n",
    "    astra_vstore = create_astra_vstore(embedding=openai_embedding, collection_name=collection_name)\n",
    "    logging.info(\"created vstore\")\n",
    "    astra_vstore.clear()\n",
    "    docs = load_pdf(file_path)\n",
    "    logging.info(\"loaded pdf\")\n",
    "    chunks = chunk_docs(docs, chunks_per_page=3, embedding=openai_embedding)\n",
    "    logging.info(\"docs chunked\")\n",
    "    astra_vstore.add_documents(chunks)\n",
    "    return astra_vstore, docs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def eval_pipeline(vstore, docs):\n",
    "    \"\"\"\n",
    "    Generate a test set using the docs and evaluate it. Output results to xlsx file.\n",
    "    \n",
    "    Args:\n",
    "        vstore (VStore): The vstore object used as a retriever.\n",
    "        docs (list): The list of documents to evaluate.\n",
    "    \n",
    "    Returns:\n",
    "        dict: A dictionary containing the averaged evaluation results.\n",
    "    \"\"\"\n",
    "    retriever = vstore.as_retriever(search_type=\"similarity_score_threshold\", search_kwargs={'score_threshold': score_threshold})\n",
    "    test_set = make_test_set(docs, test_size=10, llm=llm, embedding=openai_embedding)\n",
    "    test_set = add_llm_answers_to_dataset(test_set, retriever, llm)\n",
    "    averaged_results = evaluate_dataset(test_set)\n",
    "    logging.info(\"The averaged results are: {}\".format(averaged_results))\n",
    "    return averaged_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:astrapy.db:ASTRA_DB_KEYSPACE is not set. Defaulting to 'default_keyspace'\n",
      "INFO:astrapy.db:ASTRA_DB_KEYSPACE is not set. Defaulting to 'default_keyspace'\n",
      "INFO:httpx:HTTP Request: POST https://1da26773-d99d-4e8f-a3ce-5d34f0405b09-us-east-2.apps.astra.datastax.com/api/json/v1/default_keyspace \"HTTP/1.1 200 OK\"\n",
      "INFO:root:created vstore\n",
      "INFO:httpx:HTTP Request: POST https://1da26773-d99d-4e8f-a3ce-5d34f0405b09-us-east-2.apps.astra.datastax.com/api/json/v1/default_keyspace/aha_guidelines \"HTTP/1.1 200 OK\"\n",
      "INFO:root:loaded pdf\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:root:docs chunked\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://1da26773-d99d-4e8f-a3ce-5d34f0405b09-us-east-2.apps.astra.datastax.com/api/json/v1/default_keyspace/aha_guidelines \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://1da26773-d99d-4e8f-a3ce-5d34f0405b09-us-east-2.apps.astra.datastax.com/api/json/v1/default_keyspace/aha_guidelines \"HTTP/1.1 200 OK\"\n"
     ]
    }
   ],
   "source": [
    "vstore, docs = vstore_pipeline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://1da26773-d99d-4e8f-a3ce-5d34f0405b09-us-east-2.apps.astra.datastax.com/api/json/v1/default_keyspace/aha_guidelines \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://1da26773-d99d-4e8f-a3ce-5d34f0405b09-us-east-2.apps.astra.datastax.com/api/json/v1/default_keyspace/aha_guidelines \"HTTP/1.1 200 OK\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        Question: <br>If my <span style='background-color: yellow'>patient</span> has hypertrophic <span style='background-color: yellow'>cardiomyopathy</span> what else are they <span style='background-color: yellow'>susceptible</span> to?<br><br>\n",
       "        Answer: <br><span style='background-color: lightblue'>Myocardial</span> <span style='background-color: lightblue'>ischemia,</span> <span style='background-color: lightblue'>mitral</span> regurgitation, <span style='background-color: lightblue'>arrhythmias,</span> <span style='background-color: lightblue'>autonomic</span> <span style='background-color: lightblue'>dysfunction,</span> and <span style='background-color: lightblue'>dynamic</span> <span style='background-color: lightblue'>left</span> <span style='background-color: lightblue'>ventricular</span> <span style='background-color: lightblue'>outflow</span> <span style='background-color: lightblue'>tract</span> obstruction.\n",
       "        <br><br>\n",
       "        \n",
       "            Context 1 (score: 0.789)<br> \n",
       "            Source aha_hcm_guidance_2020_Part1.pdf page 8<br>\n",
       "            Content:<br>DOI: 10.1161/CIR.0000000000000937\n",
       "e566\n",
       "CLINICAL STATEMENTS  \n",
       "AND GUIDELINES\n",
       "pathogenic genetic variant. A substantial proportion of \n",
       "<span style='background-color: yellow'>patient</span>s with HCM are currently without any evidence of \n",
       "a genetic etiology to their disease, including a subgroup \n",
       "(up to 40% of <span style='background-color: yellow'>patient</span>s in 1 study) who also have no \n",
       "other affected family members (ie, “non-familial” \n",
       "HCM).2 These observations suggest that other novel \n",
       "pathophysiologic mechanisms may be responsible or \n",
       "contribute to phenotypic expression in these affected \n",
       "<span style='background-color: yellow'>patient</span>s with HCM. Among <span style='background-color: yellow'>patient</span>s with HCM and a pathogenic sar-\n",
       "comeric gene variant, the 2 most common genes are \n",
       "beta myosin heavy chain 7 (MYH7) and myosin-binding \n",
       "protein C3 (MYBPC3), identified in 70% of variant-pos-\n",
       "itive <span style='background-color: yellow'>patient</span>s, while other genes (TNNI3, TNNT2, TPM1, \n",
       "MYL2, MYL3, ACTC1) each account for a small propor-\n",
       "tion of <span style='background-color: yellow'>patient</span>s (1% to 5%). Within these genes, >1500 \n",
       "variants have been recognized, the majority of which \n",
       "are “private” (unique to the individual family). Each off-\n",
       "spring of an affected family member has a 50% chance \n",
       "of inheriting the variant.3 Although the likelihood of de-\n",
       "veloping clinical HCM is high in family members with a \n",
       "pathogenic variant, the age at which disease expression \n",
       "occurs in a given individual is variable. The precise mechanisms by which sarcomere vari-\n",
       "ants result in the clinical phenotype have not been fully \n",
       "elucidated. Mutant sarcomere genes trigger myocardial \n",
       "changes, leading to hypertrophy and fibrosis, which ulti-\n",
       "mately results in a small, stiff ventricle with impaired sys-\n",
       "tolic and diastolic performance despite a preserved LVEF. Similarly, abnormal sarcomeric proteins may not be solely \n",
       "responsible for all of the clinical characteristics observed \n",
       "in <span style='background-color: yellow'>patient</span>s with HCM. Diverse disease features includ-\n",
       "ing abnormal intramural coronary arteries responsible for \n",
       "small vessel <span style='background-color: lightblue'>ischemia,</span> elongated <span style='background-color: lightblue'>mitral</span> valve leaflets, and \n",
       "congenital anomalies of the sub-<span style='background-color: lightblue'>mitral</span> valve apparatus, \n",
       "which are widely recognized components of the HCM \n",
       "phenotype, appear to have no known direct association \n",
       "with sarcomere variants. 2.5. Natural History/Clinical Course\n",
       "Although HCM can be compatible with normal life ex-\n",
       "pectancy without limiting symptoms or the need for ma-\n",
       "jor treatments in most <span style='background-color: yellow'>patient</span>s, other <span style='background-color: yellow'>patient</span>s can expe-\n",
       "rience significant consequences that are attributable to \n",
       "the disease. To this point, there is increasing recognition \n",
       "of <span style='background-color: yellow'>patient</span>s with HCM identified clinically at advanced \n",
       "ages of >60 years with little to no disability. Yet, a multi-\n",
       "center registry report has suggested that the lifelong risk \n",
       "of adverse events (eg, mortality, HF, stroke, <span style='background-color: lightblue'>ventricular</span> \n",
       "arrhythmia, AF) caused by HCM may be greater among \n",
       "<span style='background-color: yellow'>patient</span>s with pathogenic sarcomeric gene variants or \n",
       "those diagnosed early in life.1 The large number and di-\n",
       "versity of the HCM-associated variants does not allow \n",
       "the specific genotype to be used to inform the antici-\n",
       "pated outcomes in individual <span style='background-color: yellow'>patient</span>s. Among referral-based cohorts of <span style='background-color: yellow'>patient</span>s with HCM, \n",
       "30% to 40% will experience adverse events, including: 1) \n",
       "sudden death events; 2) progressive limiting symptoms be-\n",
       "cause of LVOTO or diastolic dysfunction; 3) HF symptoms \n",
       "associated with systolic dysfunction; and 4) AF with risk \n",
       "of thromboembolic stroke. Nevertheless, studies reporting \n",
       "relatively long-term HCM <span style='background-color: yellow'>patient</span> outcomes have demon-\n",
       "strated that for <span style='background-color: yellow'>patient</span>s at risk for, or who develop one \n",
       "of these, disease-related complications, the application of \n",
       "contemporary cardiovascular therapies and interventions \n",
       "has lowered HCM mortality rates to <1.0%/year.2,3 One \n",
       "of the major treatment initiatives responsible for lowering \n",
       "mortality has been the evolution of SCD risk stratification \n",
       "strategies based on a number of major noninvasive risk \n",
       "markers, which can identify adult <span style='background-color: yellow'>patient</span>s with HCM at \n",
       "greatest risk for sudden death who are then candidates \n",
       "for implantable cardioverter-defibrillator (ICD) placement. The decrease in sudden death rates in HCM appears now \n",
       "to have shifted focus to heart failure (HF) as the predomi-\n",
       "nant cause of disease-related morbidity and mortality and, \n",
       "therefore, greatest unmet treatment need in adults. 3. PATHOPHYSIOLOGY\n",
       "The pathophysiology of HCM consists of <span style='background-color: lightblue'>dynamic</span> LVO-\n",
       "TO, <span style='background-color: lightblue'>mitral</span> regurgitation (MR), diastolic <span style='background-color: lightblue'>dysfunction,</span> myo-\n",
       "cardial <span style='background-color: lightblue'>ischemia,</span> <span style='background-color: lightblue'>arrhythmias,</span> and <span style='background-color: lightblue'>autonomic</span> dysfunc-\n",
       "tion. For a given <span style='background-color: yellow'>patient</span> with HCM, the clinical outcome \n",
       "may be dominated by one of these components or may \n",
       "be the result of a complex interplay. Thus, it is prudent \n",
       "to consider the potential presence of such abnormalities \n",
       "in a comprehensive clinical evaluation and address their \n",
       "impact in the management of these <span style='background-color: yellow'>patient</span>s.<br><br>\n",
       "            \n",
       "            Context 2 (score: 0.785)<br> \n",
       "            Source aha_hcm_guidance_2020_Part1.pdf page 2<br>\n",
       "            Content:<br>Ommen et al\b\n",
       "2020 AHA/ACC Guideline for Hypertrophic Cardiomyopathy\n",
       "December 22/29, 2020\b\n",
       "Circulation. 2020;142:e558–e631. DOI: 10.1161/CIR.0000000000000937\n",
       "e560\n",
       "CLINICAL STATEMENTS  \n",
       "AND GUIDELINES\n",
       "As screening recommendations for family mem-\n",
       "bers hinge on the pathogenicity of any detected \n",
       "variants, the reported pathogenicity should be \n",
       "reconfirmed every 2 to 3 years. 4. Optimal care for <span style='background-color: yellow'>patient</span>s with HCM requires cardiac \n",
       "imaging to confirm the diagnosis, characterize the \n",
       "pathophysiology for the individual, and identify risk \n",
       "markers that may inform decisions regarding inter-\n",
       "ventions for <span style='background-color: lightblue'>left</span> <span style='background-color: lightblue'>ventricular</span> <span style='background-color: lightblue'>outflow</span> <span style='background-color: lightblue'>tract</span> obstruc-\n",
       "tion and sudden cardiac death (SCD) prevention. Echocardiography continues to be the founda-\n",
       "tional imaging modality for <span style='background-color: yellow'>patient</span>s with HCM. Cardiovascular magnetic resonance imaging will \n",
       "also be helpful in many <span style='background-color: yellow'>patient</span>s, especially those in \n",
       "whom there is diagnostic uncertainty, poor echocar-\n",
       "diographic imaging windows, or where uncertainty \n",
       "persists regarding decisions around implantable car-\n",
       "dioverter-defibrillator (ICD) placement. 5. Assessment of an individual <span style='background-color: yellow'>patient</span>’s risk for \n",
       "SCD continues to evolve as new markers emerge \n",
       "(eg, apical aneurysm, decreased <span style='background-color: lightblue'>left</span> ventricu-\n",
       "lar systolic function, and extensive gadolinium \n",
       "enhancement). In addition to a full accounting of \n",
       "an individual’s risk markers, communication with \n",
       "<span style='background-color: yellow'>patient</span>s regarding not just the presence of risk \n",
       "markers but also the magnitude of their individu-\n",
       "alized risk is key. This enables the informed <span style='background-color: yellow'>patient</span> \n",
       "to fully participate in the decision-making regard-\n",
       "ing ICD placement, which incorporates their own \n",
       "level of risk tolerance and treatment goals.<br><br>\n",
       "            \n",
       "            Context 3 (score: 0.782)<br> \n",
       "            Source aha_hcm_guidance_2020_Part1.pdf page 9<br>\n",
       "            Content:<br>3.3. Mitral Regurgitation\n",
       "Mitral regurgitation (MR) can occur secondarily from \n",
       "LVOTO or from primary leaflet abnormalities and \n",
       "contributes to symptoms of dyspnea. In MR caused \n",
       "by LVOTO, SAM of the <span style='background-color: lightblue'>mitral</span> valve leads to loss of \n",
       "leaflet coaptation, and the jet is predominantly mid-\n",
       "to-late systolic and posterior or lateral in orientation.1 \n",
       "A posteriorly directed jet of MR in obstructive HCM \n",
       "correlates with SAM of the <span style='background-color: lightblue'>mitral</span> valve as the \n",
       "underlying pathophysiologic mechanism. However, \n",
       "central and anterior jets may also result from SAM of \n",
       "the <span style='background-color: lightblue'>mitral</span> valve (ie, these jets do not reliably predict \n",
       "primary <span style='background-color: lightblue'>mitral</span> leaflet abnormalities), and caution \n",
       "is necessary in using the jet direction of MR on \n",
       "preoperative transthoracic echocardiogram (TTE) to \n",
       "guide the decision for concomitant <span style='background-color: lightblue'>mitral</span> valve surgery \n",
       "during septal myectomy for HCM. Factors that affect \n",
       "the severity of LVOTO also may affect the degree of \n",
       "MR. Thus, significant MR may not be evident without \n",
       "provocation for LVOTO and SAM of the <span style='background-color: lightblue'>mitral</span> valve. Primary abnormalities of the <span style='background-color: lightblue'>mitral</span> valve and its \n",
       "apparatus are also common, including excessive \n",
       "leaflet length, anomalous papillary muscle insertion, \n",
       "and anteriorly displaced papillary muscles.2–4 In some \n",
       "<span style='background-color: yellow'>patient</span>s, these primary <span style='background-color: lightblue'>mitral</span> valve abnormalities may \n",
       "be the principal cause of symptoms. For <span style='background-color: yellow'>patient</span>s in \n",
       "whom SRT is being contemplated, close examination \n",
       "for <span style='background-color: lightblue'>mitral</span> valve abnormalities should be performed to \n",
       "determine the optimal invasive approach.5,6\n",
       "3.4. <span style='background-color: lightblue'>Myocardial</span> Ischemia\n",
       "Patients with HCM are <span style='background-color: yellow'>susceptible</span> to myocardial \n",
       "ischemia attributable to a mismatch between myocardial \n",
       "oxygen supply and demand. <span style='background-color: lightblue'>Myocardial</span> hypertrophy, \n",
       "microvascular dysfunction with impaired coronary \n",
       "flow reserve, and medial hypertrophy of the intramural \n",
       "arterioles and their reduced density are common \n",
       "findings.1,2 These abnormalities are worsened by the \n",
       "presence of hyper<span style='background-color: lightblue'>dynamic</span> systolic function and LVOTO \n",
       "with high intracavitary pressures.3,4 Blunted coronary \n",
       "flow reserve occurs even without epicardial stenosis, \n",
       "although the presence of concomitant severe coronary \n",
       "atherosclerosis exacerbates mismatch and is associated \n",
       "with a poorer prognosis.5 The presence of myocardial \n",
       "ischemia may lead to infarction, which may be evident as \n",
       "LGE on CMR imaging.6 Apical myocardial ischemia and \n",
       "infarction (with or without mid<span style='background-color: lightblue'>ventricular</span> obstruction) \n",
       "Downloaded from http://ahajournals.org by on April 30, 2024\n",
       "<br><br>\n",
       "            \n",
       "            Context 4 (score: 0.777)<br> \n",
       "            Source aha_hcm_guidance_2020_Part1.pdf page 7<br>\n",
       "            Content:<br>Ommen et al\b\n",
       "2020 AHA/ACC Guideline for Hypertrophic Cardiomyopathy\n",
       "Circulation. 2020;142:e558–e631. DOI: 10.1161/CIR.0000000000000937\b\n",
       "December 22/29, 2020\n",
       "e565\n",
       "CLINICAL STATEMENTS  \n",
       "AND GUIDELINES\n",
       "writing committee recommends the term HCM (with or \n",
       "without <span style='background-color: lightblue'>outflow</span> <span style='background-color: lightblue'>tract</span> obstruction). In some areas, the use of HCM to describe the \n",
       "increased LV wall thickness associated with systemic \n",
       "disorders or secondary causes of LV hypertrophy (LVH) \n",
       "can lead to confusion. Systemic disorders include \n",
       "various metabolic and multiorgan syndromes such as \n",
       "RASopathies (variants in several genes involved in RAS-\n",
       "MAPK signaling), mitochondrial myopathies, glycogen/\n",
       "lysosomal storage diseases in children, and Fabry, amyloid, \n",
       "sarcoid, hemochromatosis, Danon <span style='background-color: yellow'>cardiomyopathy</span> in \n",
       "adults. In these diseases, although the magnitude and \n",
       "distribution of increased LV wall thickness can be similar \n",
       "to that of isolated HCM caused by variants in sarcomeric \n",
       "genes, the pathophysiologic mechanisms responsible for \n",
       "hypertrophy, natural history, and treatment strategies \n",
       "are not the same.1–5 For these reasons, other cardiac or \n",
       "systemic diseases capable of producing LVH should not \n",
       "be labeled as HCM and will not be addressed in this \n",
       "document. In addition, other scenarios can arise that present diag-\n",
       "nostic challenges, including conditions that produce sec-\n",
       "ondary LVH, which can also overlap phenotypically with \n",
       "HCM, including remodeling secondary to athletic train-\n",
       "ing (ie, “athletes heart”) as well as morphologic changes \n",
       "related to long-standing systemic hypertension (ie, hy-\n",
       "pertensive <span style='background-color: yellow'>cardiomyopathy</span>). Similarly, hemo<span style='background-color: lightblue'>dynamic</span> ob-\n",
       "struction caused by <span style='background-color: lightblue'>left</span>-sided obstructive lesions (valvular \n",
       "or subvalvular stenosis) or obstruction after antero-apical \n",
       "infarction and stress <span style='background-color: yellow'>cardiomyopathy</span> can cause diagnos-\n",
       "tic dilemmas.6,7 Although HCM cannot be definitely ex-\n",
       "cluded in such situations, a number of clinical markers \n",
       "and testing strategies can be used to help differentiate \n",
       "between HCM and conditions of physiologic LVH. 2.3. Definition, Clinical Diagnosis, and \n",
       "Phenotype\n",
       "For the purposes of this guideline, we have considered \n",
       "the clinical definition of HCM as a disease state in which \n",
       "morphologic expression is confined solely to the heart. It is characterized predominantly by LVH in the absence \n",
       "of another cardiac, systemic, or metabolic disease ca-\n",
       "pable of producing the magnitude of hypertrophy evi-\n",
       "dent in a given <span style='background-color: yellow'>patient</span> and for which a disease-causing \n",
       "sarcomere (or sarcomere-related) variant is identified, \n",
       "or genetic etiology remains unresolved. A clinical diagnosis of HCM in adult <span style='background-color: yellow'>patient</span>s can \n",
       "therefore be established by imaging (Section 6.1), \n",
       "with 2D echocardiography or cardiovascular magnetic \n",
       "resonance (CMR) showing a maximal end-diastolic wall \n",
       "thickness of ≥15 mm anywhere in the <span style='background-color: lightblue'>left</span> ventricle, \n",
       "in the absence of another cause of hypertrophy in \n",
       "adults.1–4 More limited hypertrophy (13–14 mm) can be \n",
       "diagnostic when present in family members of a <span style='background-color: yellow'>patient</span> \n",
       "with HCM or in conjunction with a positive genetic test. For children, the diagnostic criteria are confounded by \n",
       "needing to adjust for body size and growth.<br><br>\n",
       "            \n",
       "        \n",
       "    "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "retriever = vstore.as_retriever(search_type = \"similarity_score_threshold\", search_kwargs = {'score_threshold': score_threshold})\n",
    "question = \"If my patient has hypertrophic cardiomyopathy what else are they susceptible to?\"\n",
    "answer, context = perform_rag(question, llm, retriever)\n",
    "display(HTML(format_answer(question, answer, vstore, score_threshold)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b98abc49a4474f03899f46f8cd30bf86",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "embedding nodes:   0%|          | 0/38 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:ragas.testset.docstore:Document [ID: 7ba7bbc6-8bff-460c-9c5c-30c1badd53c8] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 7ba7bbc6-8bff-460c-9c5c-30c1badd53c8] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 9e3d6883-8cb2-4404-a14b-6b0e6f5c1915] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 9e3d6883-8cb2-4404-a14b-6b0e6f5c1915] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: b4ac8eb2-e6f7-42a2-b6bb-85d6a705bce5] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: b4ac8eb2-e6f7-42a2-b6bb-85d6a705bce5] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: b00020d7-75ff-4167-9012-f6380642c845] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: b00020d7-75ff-4167-9012-f6380642c845] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: ac480447-aba4-4ba4-bd33-05235972c419] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: ac480447-aba4-4ba4-bd33-05235972c419] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 813b835a-c8a3-479d-9eda-404b394152fd] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 813b835a-c8a3-479d-9eda-404b394152fd] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 9d03ecd3-1c43-48e1-8d65-31d6f79ed064] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 9d03ecd3-1c43-48e1-8d65-31d6f79ed064] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: eba7055d-24ec-4c3e-8502-4896ee605640] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: eba7055d-24ec-4c3e-8502-4896ee605640] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: b9239a73-0b03-4c3b-8a48-c29fe6bbd40e] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: b9239a73-0b03-4c3b-8a48-c29fe6bbd40e] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 18e46055-cd8e-44af-8c14-7d38bac0ebc9] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 18e46055-cd8e-44af-8c14-7d38bac0ebc9] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 1281b3b4-bf24-4b5a-8f57-b6137474b3f3] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 1281b3b4-bf24-4b5a-8f57-b6137474b3f3] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 62e9c60f-b7c0-4e25-bd7e-52bdee448ec7] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 62e9c60f-b7c0-4e25-bd7e-52bdee448ec7] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 82f58eab-c9fa-4171-9612-ebf97dcba661] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 82f58eab-c9fa-4171-9612-ebf97dcba661] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 2e2850c9-20f2-4136-8370-726d81a9159b] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 2e2850c9-20f2-4136-8370-726d81a9159b] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: ae7033b1-eb46-4b8e-949e-acdb8afab1dd] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: ae7033b1-eb46-4b8e-949e-acdb8afab1dd] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: e1435f40-7e88-4bd4-b405-97c1fdb79f95] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: e1435f40-7e88-4bd4-b405-97c1fdb79f95] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 27299fa0-c4be-411a-ade8-38495f1860fa] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 27299fa0-c4be-411a-ade8-38495f1860fa] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 4936609f-4877-4f09-a2e5-50f103fa589e] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 4936609f-4877-4f09-a2e5-50f103fa589e] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 38392e96-ccea-49a5-83df-31cbb450d0d4] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 38392e96-ccea-49a5-83df-31cbb450d0d4] has no filename, using `doc_id` instead\n",
      "WARNING:ragas.testset.docstore:Filename and doc_id are the same for all nodes.\n",
      "INFO:ragas.testset.docstore:Document [ID: 7ba7bbc6-8bff-460c-9c5c-30c1badd53c8] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 9e3d6883-8cb2-4404-a14b-6b0e6f5c1915] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 9e3d6883-8cb2-4404-a14b-6b0e6f5c1915] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: b4ac8eb2-e6f7-42a2-b6bb-85d6a705bce5] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: b4ac8eb2-e6f7-42a2-b6bb-85d6a705bce5] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: b00020d7-75ff-4167-9012-f6380642c845] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: b00020d7-75ff-4167-9012-f6380642c845] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: ac480447-aba4-4ba4-bd33-05235972c419] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: ac480447-aba4-4ba4-bd33-05235972c419] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 813b835a-c8a3-479d-9eda-404b394152fd] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 813b835a-c8a3-479d-9eda-404b394152fd] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 9d03ecd3-1c43-48e1-8d65-31d6f79ed064] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 9d03ecd3-1c43-48e1-8d65-31d6f79ed064] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: eba7055d-24ec-4c3e-8502-4896ee605640] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: eba7055d-24ec-4c3e-8502-4896ee605640] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: b9239a73-0b03-4c3b-8a48-c29fe6bbd40e] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: b9239a73-0b03-4c3b-8a48-c29fe6bbd40e] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 18e46055-cd8e-44af-8c14-7d38bac0ebc9] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 18e46055-cd8e-44af-8c14-7d38bac0ebc9] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 1281b3b4-bf24-4b5a-8f57-b6137474b3f3] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 1281b3b4-bf24-4b5a-8f57-b6137474b3f3] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 62e9c60f-b7c0-4e25-bd7e-52bdee448ec7] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 62e9c60f-b7c0-4e25-bd7e-52bdee448ec7] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 82f58eab-c9fa-4171-9612-ebf97dcba661] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 82f58eab-c9fa-4171-9612-ebf97dcba661] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 2e2850c9-20f2-4136-8370-726d81a9159b] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 2e2850c9-20f2-4136-8370-726d81a9159b] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: ae7033b1-eb46-4b8e-949e-acdb8afab1dd] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: ae7033b1-eb46-4b8e-949e-acdb8afab1dd] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: e1435f40-7e88-4bd4-b405-97c1fdb79f95] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: e1435f40-7e88-4bd4-b405-97c1fdb79f95] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 27299fa0-c4be-411a-ade8-38495f1860fa] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 27299fa0-c4be-411a-ade8-38495f1860fa] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 4936609f-4877-4f09-a2e5-50f103fa589e] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 4936609f-4877-4f09-a2e5-50f103fa589e] has no filename, using `doc_id` instead\n",
      "INFO:ragas.testset.docstore:Document [ID: 38392e96-ccea-49a5-83df-31cbb450d0d4] has no filename, using `doc_id` instead\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fdddb21475904aec8576d71af6fcf585",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Generating:   0%|          | 0/10 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:asyncio:Task exception was never retrieved\n",
      "future: <Task finished name='Task-612' coro=<AsyncClient.aclose() done, defined at c:\\Users\\jmachalek\\.virtualenvs\\mini_project-kSexw87t\\Lib\\site-packages\\httpx\\_client.py:2011> exception=RuntimeError('Event loop is closed')>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\jmachalek\\.virtualenvs\\mini_project-kSexw87t\\Lib\\site-packages\\httpx\\_client.py\", line 2018, in aclose\n",
      "    await self._transport.aclose()\n",
      "  File \"c:\\Users\\jmachalek\\.virtualenvs\\mini_project-kSexw87t\\Lib\\site-packages\\httpx\\_transports\\default.py\", line 385, in aclose\n",
      "    await self._pool.aclose()\n",
      "  File \"c:\\Users\\jmachalek\\.virtualenvs\\mini_project-kSexw87t\\Lib\\site-packages\\httpcore\\_async\\connection_pool.py\", line 313, in aclose\n",
      "    await self._close_connections(closing_connections)\n",
      "  File \"c:\\Users\\jmachalek\\.virtualenvs\\mini_project-kSexw87t\\Lib\\site-packages\\httpcore\\_async\\connection_pool.py\", line 305, in _close_connections\n",
      "    await connection.aclose()\n",
      "  File \"c:\\Users\\jmachalek\\.virtualenvs\\mini_project-kSexw87t\\Lib\\site-packages\\httpcore\\_async\\connection.py\", line 171, in aclose\n",
      "    await self._connection.aclose()\n",
      "  File \"c:\\Users\\jmachalek\\.virtualenvs\\mini_project-kSexw87t\\Lib\\site-packages\\httpcore\\_async\\http11.py\", line 265, in aclose\n",
      "    await self._network_stream.aclose()\n",
      "  File \"c:\\Users\\jmachalek\\.virtualenvs\\mini_project-kSexw87t\\Lib\\site-packages\\httpcore\\_backends\\anyio.py\", line 55, in aclose\n",
      "    await self._stream.aclose()\n",
      "  File \"c:\\Users\\jmachalek\\.virtualenvs\\mini_project-kSexw87t\\Lib\\site-packages\\anyio\\streams\\tls.py\", line 202, in aclose\n",
      "    await self.transport_stream.aclose()\n",
      "  File \"c:\\Users\\jmachalek\\.virtualenvs\\mini_project-kSexw87t\\Lib\\site-packages\\anyio\\_backends\\_asyncio.py\", line 1191, in aclose\n",
      "    self._transport.close()\n",
      "  File \"C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.12_3.12.1008.0_x64__qbz5n2kfra8p0\\Lib\\asyncio\\selector_events.py\", line 1210, in close\n",
      "    super().close()\n",
      "  File \"C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.12_3.12.1008.0_x64__qbz5n2kfra8p0\\Lib\\asyncio\\selector_events.py\", line 875, in close\n",
      "    self._loop.call_soon(self._call_connection_lost, None)\n",
      "  File \"C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.12_3.12.1008.0_x64__qbz5n2kfra8p0\\Lib\\asyncio\\base_events.py\", line 795, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.12_3.12.1008.0_x64__qbz5n2kfra8p0\\Lib\\asyncio\\base_events.py\", line 541, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:ragas.testset.evolutions:seed question generated: What role do pathogenic sarcomere variants play in the development of hypertrophic cardiomyopathy (HCM)?\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:ragas.testset.evolutions:seed question generated: What is the purpose of the new guideline commissioned by Ommen et al in the 2020 AHA/ACC Guideline for Hypertrophic Cardiomyopathy?\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:ragas.testset.evolutions:seed question generated: How does a posteriorly directed jet of mitral regurgitation in obstructive HCM correlate with SAM of the mitral valve as the underlying pathophysiologic mechanism?\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:ragas.testset.evolutions:seed question generated: How has the focus shifted from sudden death risk to heart failure as the predominant cause of morbidity and mortality in patients with hypertrophic cardiomyopathy (HCM)?\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:ragas.testset.evolutions:seed question generated: How should clinical judgment be used in conjunction with guideline recommendations in patient care?\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:ragas.testset.evolutions:seed question generated: What is the role of hypertrophy in the development of Hypertrophic Cardiomyopathy (HCM)?\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:ragas.testset.evolutions:seed question generated: What are the two most common genes associated with sarcomeric gene variants in patients with hypertrophic cardiomyopathy (HCM)?\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:ragas.testset.evolutions:seed question generated: What are the key recommendations for optimal care and management of patients with Hypertrophic Cardiomyopathy according to the 2020 AHA/ACC Guideline?\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:ragas.testset.evolutions:seed question generated: What does SCD stand for in the context provided?\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:ragas.testset.evolutions:seed question generated: What is the role of LVOT obstruction in hypertrophic cardiomyopathy according to the 2020 AHA/ACC Guideline?\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 429 Too Many Requests\"\n",
      "INFO:openai._base_client:Retrying request to /chat/completions in 0.149000 seconds\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 429 Too Many Requests\"\n",
      "INFO:openai._base_client:Retrying request to /chat/completions in 0.571000 seconds\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 429 Too Many Requests\"\n",
      "INFO:openai._base_client:Retrying request to /chat/completions in 0.716000 seconds\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 429 Too Many Requests\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 429 Too Many Requests\"\n",
      "INFO:openai._base_client:Retrying request to /chat/completions in 0.979000 seconds\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 429 Too Many Requests\"\n",
      "INFO:openai._base_client:Retrying request to /chat/completions in 0.314000 seconds\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://1da26773-d99d-4e8f-a3ce-5d34f0405b09-us-east-2.apps.astra.datastax.com/api/json/v1/default_keyspace/aha_guidelines \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://1da26773-d99d-4e8f-a3ce-5d34f0405b09-us-east-2.apps.astra.datastax.com/api/json/v1/default_keyspace/aha_guidelines \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://1da26773-d99d-4e8f-a3ce-5d34f0405b09-us-east-2.apps.astra.datastax.com/api/json/v1/default_keyspace/aha_guidelines \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://1da26773-d99d-4e8f-a3ce-5d34f0405b09-us-east-2.apps.astra.datastax.com/api/json/v1/default_keyspace/aha_guidelines \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://1da26773-d99d-4e8f-a3ce-5d34f0405b09-us-east-2.apps.astra.datastax.com/api/json/v1/default_keyspace/aha_guidelines \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://1da26773-d99d-4e8f-a3ce-5d34f0405b09-us-east-2.apps.astra.datastax.com/api/json/v1/default_keyspace/aha_guidelines \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://1da26773-d99d-4e8f-a3ce-5d34f0405b09-us-east-2.apps.astra.datastax.com/api/json/v1/default_keyspace/aha_guidelines \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://1da26773-d99d-4e8f-a3ce-5d34f0405b09-us-east-2.apps.astra.datastax.com/api/json/v1/default_keyspace/aha_guidelines \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://1da26773-d99d-4e8f-a3ce-5d34f0405b09-us-east-2.apps.astra.datastax.com/api/json/v1/default_keyspace/aha_guidelines \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://1da26773-d99d-4e8f-a3ce-5d34f0405b09-us-east-2.apps.astra.datastax.com/api/json/v1/default_keyspace/aha_guidelines \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1bcca381efb74f84b4adab860132fa22",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Evaluating:   0%|          | 0/30 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:httpx:HTTP Request: POST https://api.openai.com/v1/embeddings \"HTTP/1.1 200 OK\"\n",
      "INFO:root:The averaged results are: {'context_precision': 1.0000, 'context_recall': 1.0000, 'answer_correctness': 0.5652}\n"
     ]
    }
   ],
   "source": [
    "avg_result = eval_pipeline(vstore, docs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'context_precision': 1.0000, 'context_recall': 1.0000, 'answer_correctness': 0.7652}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(avg_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#vstore.delete_collection()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mini_project-kSexw87t",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
